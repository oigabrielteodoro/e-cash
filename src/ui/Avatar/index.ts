export * from './Avatar'
export * from './NotFoundAvatar'
