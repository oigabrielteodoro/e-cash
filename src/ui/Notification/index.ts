export const NOTIFICATION_EVENT_NAME = 'notification'
export const EXPIRES_NOTIFICATION_IN_SECONDS = 1000 * 3 // 3 seconds

export * from './Notification'
export * from './NotificationContainer'
