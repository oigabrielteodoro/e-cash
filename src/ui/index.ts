export * from './Style'
export * from './Input'
export * from './Button'
export * from './PasswordInput'
export * from './Toast'
export * from './Logo'
export * from './SideBar'
export * from './AppLayout'
export * from './Tooltip'
export * from './Avatar'
export * from './ShimmerEffect'
export * from './Popover'
export * from './Grid'
