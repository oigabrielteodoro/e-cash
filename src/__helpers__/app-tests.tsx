import React, { useEffect, ReactElement, ReactNode } from 'react'
import { render as rtlRender } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import {
  DefaultOptions,
  QueryCache,
  QueryClient,
  QueryClientProvider,
} from 'react-query'
import { MemoryRouter, Routes, Route } from 'react-router-dom'
import { generateId } from 'lib'
import nock from 'nock'

import { pipe } from 'fp-ts/function'
import { TaskEither, map, mapLeft } from 'fp-ts/TaskEither'

import { AppLayout, NotificationContainer } from 'ui'
import { clearToken, queryConfigDefault, setToken } from 'client'

type Options = {
  initialRoute?: string
  initialRoutePath?: string
  authenticated?: boolean
  queryClient?: QueryClient
  routePaths?: string[]
}

type Props = {
  children?: ReactNode
  client?: QueryClient
}

type Callback = (result: unknown) => unknown
type MapAll = (
  fn: Callback,
) => (data: TaskEither<unknown, unknown>) => TaskEither<unknown, unknown>

const queryCache = new QueryCache()
const defaultQueryClient = new QueryClient({
  defaultOptions: queryConfigDefault as DefaultOptions,
  queryCache: queryCache,
})

export function ReactQueryWrapper({
  children,
  client = defaultQueryClient,
}: Props) {
  return <QueryClientProvider client={client}>{children}</QueryClientProvider>
}

export function ReactRouterWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
}: Options & Pick<Props, 'children'>) {
  function MockedPage({ title }: { title: string }) {
    if (authenticated) {
      return (
        <AppLayout>
          <AppLayout.Content>Page {title}</AppLayout.Content>
        </AppLayout>
      )
    }

    return <h1>Page {title}</h1>
  }

  return (
    <MemoryRouter initialEntries={[initialRoute]}>
      <Routes>
        <Route path={initialRoute} element={children} />
        {routePaths?.map((routePath) => (
          <Route
            key={routePath}
            path={routePath}
            element={<MockedPage title={routePath} />}
          />
        ))}
      </Routes>
    </MemoryRouter>
  )
}

export function ReactHookWrapper({
  initialRoute = '/',
  children,
  routePaths,
  authenticated = true,
  client = defaultQueryClient,
}: Options & Props) {
  return (
    <ReactQueryWrapper client={client}>
      <ReactRouterWrapper
        initialRoute={initialRoute}
        routePaths={routePaths}
        authenticated={authenticated}
      >
        {children}
      </ReactRouterWrapper>
    </ReactQueryWrapper>
  )
}

function render(
  ui: ReactElement,
  {
    initialRoute = '/',
    queryClient = defaultQueryClient,
    authenticated = true,
    routePaths,
  }: Options = {},
) {
  function Wrapper({ children }: Props) {
    useEffect(() => {
      if (authenticated) {
        setToken(generateId())

        return () => {
          clearToken()
        }
      }
    }, [])

    return (
      <ReactQueryWrapper client={queryClient}>
        <ReactRouterWrapper
          initialRoute={initialRoute}
          routePaths={routePaths}
          authenticated={authenticated}
        >
          {children}
        </ReactRouterWrapper>

        <NotificationContainer />
      </ReactQueryWrapper>
    )
  }

  return {
    ...rtlRender(ui, {
      wrapper: Wrapper,
    }),
  }
}

export function unsafe<T>(value: unknown) {
  return value as T
}

export const mapAll: MapAll = (fn) => (data) => {
  return pipe(data, map(fn), mapLeft(fn))
}

export function getErrorMessage(errors: unknown): string {
  return Array.isArray(errors) ? errors[0].message : ''
}

export * from '@testing-library/react'
export {
  render,
  userEvent,
  nock,
  defaultQueryClient as queryClient,
  queryCache,
}
