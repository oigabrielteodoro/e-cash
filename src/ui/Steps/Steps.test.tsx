import React, { useRef } from 'react'
import { ForwardedSteps, StepRefProps, Steps } from 'ui'

import { render, screen, userEvent } from '__helpers__/app-tests'

function MockedComponent() {
  const ref = useRef<StepRefProps>(null)

  return (
    <ForwardedSteps ref={ref}>
      <Steps.Step>
        <h1>Step 1</h1>
        <button onClick={() => ref.current?.goNext()}>Go next two</button>
      </Steps.Step>
      <Steps.Step>
        <h1>Step 2</h1>
        <button onClick={() => ref.current?.goBack()}>Go back one</button>
      </Steps.Step>
      <Steps.Step>
        <h1>Step 3</h1>
        <button onClick={() => ref.current?.navigate(0)}>
          Go navigate zero
        </button>
      </Steps.Step>
    </ForwardedSteps>
  )
}

describe('Steps', () => {
  it('should be able match snapshot', () => {
    const { container } = render(<MockedComponent />)

    expect(container).toMatchSnapshot()
  })

  it('should be able render correctly', () => {
    render(<MockedComponent />)

    expect(screen.getByText('Step 1')).toBeInTheDocument()
  })

  it('should be able navigate to step two using step navigation', async () => {
    render(<MockedComponent />)

    expect(screen.getByText('Step 1')).toBeInTheDocument()

    userEvent.click(screen.getByLabelText('Step Control 2'))

    expect(await screen.findByText('Step 2')).toBeInTheDocument()
  })

  it('should be able next and back', async () => {
    render(<MockedComponent />)

    expect(screen.getByText('Step 1')).toBeInTheDocument()

    userEvent.click(
      screen.getByRole('button', {
        name: 'Go next two',
      }),
    )

    expect(await screen.findByText('Step 2')).toBeInTheDocument()

    userEvent.click(
      screen.getByRole('button', {
        name: 'Go back one',
      }),
    )

    expect(await screen.findByText('Step 1')).toBeInTheDocument()
  })

  it('should be able navigate to step three and navigate to step zero', async () => {
    render(<MockedComponent />)

    expect(screen.getByText('Step 1')).toBeInTheDocument()

    userEvent.click(screen.getByLabelText('Step Control 3'))

    expect(await screen.findByText('Step 3')).toBeInTheDocument()

    userEvent.click(
      screen.getByRole('button', {
        name: 'Go navigate zero',
      }),
    )

    expect(await screen.findByText('Step 1')).toBeInTheDocument()
  })
})
