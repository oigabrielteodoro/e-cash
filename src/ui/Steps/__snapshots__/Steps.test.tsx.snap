// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Steps should be able match snapshot 1`] = `
.c2 {
  position: relative;
}

.c1 {
  width: 100%;
  position: relative;
  background: transparent;
  border: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  cursor: default;
}

.c1::after {
  content: '';
  position: absolute;
  background: #cfd7e3;
  height: 0.1rem;
  left: 4.6rem;
  right: 1.6rem;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  -webkit-animation: kKBrSp 2s;
  animation: kKBrSp 2s;
}

.c1:last-child {
  width: 3rem;
}

.c1:last-child::after {
  display: none;
}

.c3 {
  width: 3rem;
  height: 3rem;
  display: grid;
  cursor: pointer;
  -webkit-transition: 300ms;
  transition: 300ms;
  border-radius: 50%;
  place-items: center;
  border: 0.1rem solid transparent;
  color: #a1a7ae;
  border-color: #a1a7ae;
  font-weight: 500;
  border-color: transparent;
  color: #ffffff;
  background: #4361ee;
}

.c4 {
  width: 3rem;
  height: 3rem;
  display: grid;
  cursor: pointer;
  -webkit-transition: 300ms;
  transition: 300ms;
  border-radius: 50%;
  place-items: center;
  border: 0.1rem solid transparent;
  color: #a1a7ae;
  border-color: #a1a7ae;
}

.c4:hover {
  color: #18171d;
  border-color: #18171d;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
}

.c5 {
  margin: 5rem 0 2rem;
}

<div>
  <div
    class="c0"
  >
    <button
      aria-label="Step Control 1"
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="c3"
        >
          1
        </div>
      </div>
    </button>
    <button
      aria-label="Step Control 2"
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="c4"
        >
          2
        </div>
      </div>
    </button>
    <button
      aria-label="Step Control 3"
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="c4"
        >
          3
        </div>
      </div>
    </button>
  </div>
  <div
    class="c5"
  >
    <h1>
      Step 1
    </h1>
    <button>
      Go next two
    </button>
  </div>
</div>
`;
